File: src/main/java/com/danielremsburg/jaffolding/ui/layout/GridLayout.java
package com.danielremsburg.jaffolding.ui.layout;

import com.danielremsburg.jaffolding.Component;
import com.danielremsburg.jaffolding.ui.UIComponent;

/**
 * A grid layout arranges components in a grid of rows and columns.
 * Similar to GridLayout in Swing.
 */
public class GridLayout implements Layout {
    private int rows;
    private int cols;
    private int hgap;
    private int vgap;
    
    public GridLayout(int rows, int cols) {
        this(rows, cols, 0, 0);
    }
    
    public GridLayout(int rows, int cols, int hgap, int vgap) {
        this.rows = rows;
        this.cols = cols;
        this.hgap = hgap;
        this.vgap = vgap;
    }
    
    @Override
    public void applyTo(UIComponent container) {
        container.setStyle("display", "grid");
        
        if (cols > 0) {
            container.setStyle("grid-template-columns", "repeat(" + cols + ", 1fr)");
        }
        
        if (rows > 0) {
            container.setStyle("grid-template-rows", "repeat(" + rows + ", auto)");
        }
        
        if (hgap > 0 || vgap > 0) {
            container.setStyle("gap", vgap + "px " + hgap + "px");
        }
    }
    
    @Override
    public void addLayoutComponent(Component comp) {
        // No specific positioning needed
    }
    
    @Override
    public void addLayoutComponent(Component comp, Object constraints) {
        if (constraints instanceof GridConstraints) {
            GridConstraints gc = (GridConstraints) constraints;
            
            if (comp instanceof UIComponent) {
                UIComponent uiComp = (UIComponent) comp;
                
                if (gc.gridx >= 0) {
                    uiComp.setStyle("grid-column-start", String.valueOf(gc.gridx + 1));
                }
                
                if (gc.gridy >= 0) {
                    uiComp.setStyle("grid-row-start", String.valueOf(gc.gridy + 1));
                }
                
                if (gc.gridwidth > 1) {
                    uiComp.setStyle("grid-column-end", "span " + gc.gridwidth);
                }
                
                if (gc.gridheight > 1) {
                    uiComp.setStyle("grid-row-end", "span " + gc.gridheight);
                }
            }
        }
    }
    
    @Override
    public void removeLayoutComponent(Component comp) {
        // Nothing to do
    }
    
    @Override
    public void layoutContainer(UIComponent container) {
        applyTo(container);
    }
    
    /**
     * Constraints for grid layout positioning.
     */
    public static class GridConstraints {
        public int gridx = -1;
        public int gridy = -1;
        public int gridwidth = 1;
        public int gridheight = 1;
        
        public GridConstraints() {}
        
        public GridConstraints(int gridx, int gridy) {
            this.gridx = gridx;
            this.gridy = gridy;
        }
        
        public GridConstraints(int gridx, int gridy, int gridwidth, int gridheight) {
            this.gridx = gridx;
            this.gridy = gridy;
            this.gridwidth = gridwidth;
            this.gridheight = gridheight;
        }
    }
}

File: src/main/resources/static/js/jaffolding/ui/layout/GridLayout.js
/**
 * A grid layout arranges components in a grid of rows and columns.
 * JavaScript implementation that mirrors the Java API.
 */
(function() {
  class GridLayout {
    /**
     * Creates a new GridLayout instance.
     * @param {number} rows - The number of rows
     * @param {number} cols - The number of columns
     * @param {number} [hgap=0] - The horizontal gap
     * @param {number} [vgap=0] - The vertical gap
     */
    constructor(rows, cols, hgap = 0, vgap = 0) {
      this.rows = rows;
      this.cols = cols;
      this.hgap = hgap;
      this.vgap = vgap;
    }

    /**
     * Applies the layout to a container.
     * @param {Component} container - The container component
     */
    applyTo(container) {
      container.setStyle('display', 'grid');
      
      if (this.cols > 0) {
        container.setStyle('grid-template-columns', `repeat(${this.cols}, 1fr)`);
      }
      
      if (this.rows > 0) {
        container.setStyle('grid-template-rows', `repeat(${this.rows}, auto)`);
      }
      
      if (this.hgap > 0 || this.vgap > 0) {
        container.setStyle('gap', `${this.vgap}px ${this.hgap}px`);
      }
    }

    /**
     * Adds a component to the layout.
     * @param {Component} comp - The component to add
     */
    addLayoutComponent(comp) {
      // No specific positioning needed
    }

    /**
     * Adds a component to the layout with constraints.
     * @param {Component} comp - The component to add
     * @param {Object} constraints - The layout constraints
     */
    addLayoutComponent(comp, constraints) {
      if (constraints && typeof constraints === 'object') {
        if (constraints.gridx >= 0) {
          comp.setStyle('grid-column-start', String(constraints.gridx + 1));
        }
        
        if (constraints.gridy >= 0) {
          comp.setStyle('grid-row-start', String(constraints.gridy + 1));
        }
        
        if (constraints.gridwidth > 1) {
          comp.setStyle('grid-column-end', `span ${constraints.gridwidth}`);
        }
        
        if (constraints.gridheight > 1) {
          comp.setStyle('grid-row-end', `span ${constraints.gridheight}`);
        }
      }
    }

    /**
     * Removes a component from the layout.
     * @param {Component} comp - The component to remove
     */
    removeLayoutComponent(comp) {
      // Nothing to do
    }

    /**
     * Lays out the container.
     * @param {Component} container - The container component
     */
    layoutContainer(container) {
      this.applyTo(container);
    }

    /**
     * Creates grid constraints.
     * @param {number} gridx - The grid x coordinate
     * @param {number} gridy - The grid y coordinate
     * @param {number} [gridwidth=1] - The grid width
     * @param {number} [gridheight=1] - The grid height
     * @returns {Object} - The grid constraints
     * @static
     */
    static createConstraints(gridx, gridy, gridwidth = 1, gridheight = 1) {
      return {
        gridx,
        gridy,
        gridwidth,
        gridheight
      };
    }
  }

  // Export the GridLayout class
  window.jaffolding = window.jaffolding || {};
  window.jaffolding.GridLayout = GridLayout;
})();